ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -213,11 +220,41 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
 
   @Override
   public void handlePut(RestRequest restRequest, RestResponseChannel restResponseChannel) {
+    long processingStartTime = System.currentTimeMillis();
+    long preProcessingTime = 0;
     handlePrechecks(restRequest, restResponseChannel);
-    Exception exception =
-        isUp ? new RestServiceException("PUT is not supported", RestServiceErrorCode.UnsupportedHttpMethod)
-            : new RestServiceException("AmbryBlobStorageService unavailable", RestServiceErrorCode.ServiceUnavailable);
-    submitResponse(restRequest, restResponseChannel, null, exception);
+    try {
+      logger.trace("Handling PUT request - {}", restRequest.getUri());
+      checkAvailable();
+      // TODO: make this non blocking once all handling of individual methods is moved to their own classes
+      securityService.preProcessRequest(restRequest).get();
+      String operationOrBlobId =
+          RestUtils.getOperationOrBlobIdFromUri(restRequest, null, frontendConfig.frontendPathPrefixesToRemove);
+      if (operationOrBlobId.startsWith("/")) {
+        operationOrBlobId = operationOrBlobId.substring(1);
+      }
+      if (operationOrBlobId.equalsIgnoreCase(Operations.UPDATE_TTL)) {
+        ttlUpdateHandler.handle(restRequest, restResponseChannel, (r, e) -> {
+          if (e != null && e instanceof RouterException
+              && ((RouterException) e).getErrorCode() == RouterErrorCode.BlobUpdateNotAllowed) {
+            try {
+              restResponseChannel.setHeader(Headers.ALLOW, TTL_UPDATE_REJECTED_ALLOW_HEADER_VALUE);
+            } catch (RestServiceException exp) {
+              logger.error("Exception while setting {}", Headers.ALLOW, exp);
+            }
+          }
+          submitResponse(restRequest, restResponseChannel, null, e);
+        });
+      } else {
+        submitResponse(restRequest, restResponseChannel, null,
+            new RestServiceException("Unrecognized operation: " + operationOrBlobId, RestServiceErrorCode.BadRequest));
+      }
+      preProcessingTime = System.currentTimeMillis() - processingStartTime;
+    } catch (Exception e) {
+      submitResponse(restRequest, restResponseChannel, null, extractExecutionExceptionCause(e));
+    } finally {
+      frontendMetrics.putPreProcessingTimeInMs.update(preProcessingTime);
```

Comment:
could we make one `submitResponse` call in this finally block? i.e. reassign the variable `e` in the callback logic and call `submitResponse(..., null, e)` in `finally`